core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -425,272 +425,356 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
+  final def stackless[E](cause: Cause[E]): Cause[E]             = Internal.Meta(cause, Internal.Data(true))
+  final def traced[E](cause: Cause[E], trace: ZTrace): Cause[E] = Internal.Traced(cause, trace)
 
   /**
    * Converts the specified `Cause[Option[E]]` to an `Option[Cause[E]]` by
    * recursively stripping out any failures with the error `None`.
    */
   final def sequenceCauseOption[E](c: Cause[Option[E]]): Option[Cause[E]] =
     c match {
-      case Empty                      => Some(Empty)
-      case Cause.Traced(cause, trace) => sequenceCauseOption(cause).map(Cause.Traced(_, trace))
-      case Cause.Meta(cause, data)    => sequenceCauseOption(cause).map(Cause.Meta(_, data))
-      case Cause.Interrupt(id)        => Some(Cause.Interrupt(id))
-      case d @ Cause.Die(_)           => Some(d)
-      case Cause.Fail(Some(e))        => Some(Cause.Fail(e))
-      case Cause.Fail(None)           => None
-      case Cause.Then(left, right) =>
+      case Internal.Empty                => Some(Internal.Empty)
+      case Internal.Traced(cause, trace) => sequenceCauseOption(cause).map(Internal.Traced(_, trace))
+      case Internal.Meta(cause, data)    => sequenceCauseOption(cause).map(Internal.Meta(_, data))
+      case Internal.Interrupt(id)        => Some(Internal.Interrupt(id))
+      case d @ Internal.Die(_)           => Some(d)
+      case Internal.Fail(Some(e))        => Some(Internal.Fail(e))
+      case Internal.Fail(None)           => None
+      case Internal.Then(left, right) =>
         (sequenceCauseOption(left), sequenceCauseOption(right)) match {
-          case (Some(cl), Some(cr)) => Some(Cause.Then(cl, cr))
+          case (Some(cl), Some(cr)) => Some(Internal.Then(cl, cr))
           case (None, Some(cr))     => Some(cr)
           case (Some(cl), None)     => Some(cl)
           case (None, None)         => None
         }
 
-      case Cause.Both(left, right) =>
+      case Both(left, right) =>
         (sequenceCauseOption(left), sequenceCauseOption(right)) match {
-          case (Some(cl), Some(cr)) => Some(Cause.Both(cl, cr))
+          case (Some(cl), Some(cr)) => Some(Internal.Both(cl, cr))
           case (None, Some(cr))     => Some(cr)
           case (Some(cl), None)     => Some(cl)
           case (None, None)         => None
         }
     }
 
-  private final case class Fail[E](value: E) extends Cause[E] {
-    override final def equals(that: Any): Boolean = that match {
-      case fail: Fail[_]     => value == fail.value
-      case c @ Then(_, _)    => sym(empty)(this, c)
-      case c @ Both(_, _)    => sym(empty)(this, c)
-      case traced: Traced[_] => this == traced.cause
-      case meta: Meta[_]     => this == meta.cause
-      case _                 => false
-    }
+  object Empty {
+    def unapply[E](cause: Cause[E]): Boolean =
+      cause.find {
+        case Internal.Empty        => true
+        case Internal.Fail(_)      => false
```

Comment:
It looks like you could merge all the false cases in a single one `case _: Cause[_] => false`, same for all the other ones?

Reply:
even using a simple pattern match instead of reaching for `find`.

Reply:
The reason we need this is to deal with the `Meta` case. Let's say the cause is `Internal.Meta(Internal.Fail(e))`. If we just pattern match on `Internal.Fail`  we won't match, but we want to ignore `Meta` and succeed with a match here. So then we might think okay, let's just pattern match on that. But `Meta` causes can be nested arbitrarily deeply so we need something to look through them, which is where `find` comes in.

This is also why we need to enumerate the other cases. `find` takes a partial function, so the partial function being defined for a cause means "we've found what we're looking for, stop searching" whereas it not being defined means "we haven't found it yet, keep recursing". So if we're looking for `Internal.Fail(e)` and we see `Internal.Both(l, r)` we want to stop searching immediately and fail the match. Whereas if we see `Meta(cause, _)` we want to recurse into `cause` to see if it satisfies the match.

Reply:
Lovely. I missed the addition of `Meta`.